In [1]:
pip install termcolor

Note: you may need to restart the kernel to use updated packages.


In [2]:
#Packages related to general operating system & warnings
import os 
import warnings
warnings.filterwarnings('ignore')
#Packages related to data importing, manipulation, exploratory data #analysis, data understanding
import numpy as np
import pandas as pd
from pandas import Series, DataFrame
from termcolor import colored as cl # text customization
#Packages related to data visualizaiton
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
#Setting plot sizes and type of plot
plt.rc("font", size=14)
plt.rcParams['axes.grid'] = True
plt.figure(figsize=(6,3))
plt.gray()
from matplotlib.backends.backend_pdf import PdfPages
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn import metrics
from sklearn.impute import MissingIndicator, SimpleImputer
from sklearn.preprocessing import  PolynomialFeatures, KBinsDiscretizer, FunctionTransformer
from sklearn.preprocessing import StandardScaler, MinMaxScaler, MaxAbsScaler
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, LabelBinarizer, OrdinalEncoder
import statsmodels.formula.api as smf
import statsmodels.tsa as tsa
from sklearn.linear_model import LogisticRegression, LinearRegression, ElasticNet, Lasso, Ridge
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor, export_graphviz
from sklearn.ensemble import BaggingClassifier, BaggingRegressor,RandomForestClassifier,RandomForestRegressor
from sklearn.ensemble import GradientBoostingClassifier,GradientBoostingRegressor, AdaBoostClassifier, AdaBoostRegressor
from sklearn.svm import LinearSVC, LinearSVR, SVC, SVR
from xgboost import XGBClassifier
from sklearn.metrics import f1_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

<Figure size 600x300 with 0 Axes>

In [3]:
data = pd.read_csv("creditcard.csv")
data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [5]:
data.shape

(284807, 31)

In [6]:
data.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

In [7]:
data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [8]:
Total_transactions = len(data)
#len(data) returns number of rows in dataframe
normal = len(data[data.Class == 0])
fraudulent = len(data[data.Class == 1])
fraud_percentage = round(fraudulent/normal*100, 2)
print(cl('Total number of Transactions are {}'.format(Total_transactions), attrs = ['bold']))
print(cl('Number of Normal Transactions are {}'.format(normal), attrs = ['bold']))
print(cl('Number of fraudulent Transactions are {}'.format(fraudulent), attrs = ['bold']))
print(cl('Percentage of fraud Transactions is {}'.format(fraud_percentage), attrs = ['bold']))

Total number of Transactions are 284807
Number of Normal Transactions are 284315
Number of fraudulent Transactions are 492
Percentage of fraud Transactions is 0.17


In [9]:
min(data.Amount), max(data.Amount)

(0.0, 25691.16)

In [10]:
#Scale the 'Amount' variable
sc = StandardScaler()
amount = data['Amount'].values
data['Amount'] = sc.fit_transform(amount.reshape(-1,1))

In [11]:
#Drop 'Time' variable as it is not necessary for modelling process
data.drop(['Time'], axis=1, inplace = True)

In [12]:
#Remove and duplicates
data.drop_duplicates(inplace = True)

In [13]:
data.shape

(275663, 30)

In [14]:
#Create dependent and independent variables
X = data.drop('Class', axis = 1).values
y = data['Class'].values

In [15]:
#Split our train and test data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 1)

In [16]:
#Decision Tree
DT = DecisionTreeClassifier(max_depth = 4, criterion = 'entropy')
DT.fit(X_train, y_train)
dt_yhat = DT.predict(X_test)

In [17]:
#Check accuracy of Decision Tree model
print('Accuracy score of the Decision Tree model is {}'.format(accuracy_score(y_test,dt_yhat)*100))

Accuracy score of the Decision Tree model is 99.91583957281328


In [18]:
#Checking F1 score of Decision Tree model
print('F1 score of the Decision Tree model is {}'.format(f1_score(y_test, dt_yhat)*100))

F1 score of the Decision Tree model is 75.42372881355934


In [19]:
#Checking the confusion matrix
confusion_matrix(y_test, dt_yhat, labels = [0,1])

array([[68769,    19],
       [   39,    89]], dtype=int64)

In [20]:
#K-Nearest Neighbors
n = 7
KNN = KNeighborsClassifier(n_neighbors = n)
KNN.fit(X_train, y_train)
knn_yhat = KNN.predict(X_test)

In [21]:
#Check accuracy of KNN model
print('Accuracy score of the K-Nearest Neighbors model is {}'.format(accuracy_score(y_test, knn_yhat)*100))

Accuracy score of the K-Nearest Neighbors model is 99.9288989494457


In [22]:
#Check F1-score of KNN model
print('F1 score of the K-Nearest Neighbors model is {}'.format(f1_score(y_test, knn_yhat)*100))

F1 score of the K-Nearest Neighbors model is 79.49790794979079


In [23]:
#Logistic Regression
lr = LogisticRegression()
lr.fit(X_train, y_train)
lr_yhat = lr.predict(X_test)

In [24]:
#Check accuracy of Logistic Regression Model
print('Accuracy score of the Logistic Regression model is {}'.format(accuracy_score(y_test, lr_yhat)*100))

Accuracy score of the Logistic Regression model is 99.89552498694061


In [25]:
#Check F1-score of Logistic Regression Model
print('F1 score of the Logistic Regression model is {}'.format(f1_score(y_test, lr_yhat)*100))

F1 score of the Logistic Regression model is 66.66666666666666


In [26]:
#Support Vector Machines
svm = SVC()
svm.fit(X_train, y_train)
svm_yhat = svm.predict(X_test)

In [27]:
#Check accuracy of Support Vector Machines model
print('Accuracy score of the Support Vector Machines model is {}'.format(accuracy_score(y_test, svm_yhat)*100))

Accuracy score of the Support Vector Machines model is 99.9318010331418


In [28]:
#Random Forest
rf = RandomForestClassifier(max_depth = 4)
rf.fit(X_train, y_train)
rf_yhat = rf.predict(X_test)

In [29]:
#Check accuracy of Random Forest model
print('Accuracy score of the Random Forest model is {}'.format(accuracy_score(y_test, rf_yhat)*100))

Accuracy score of the Random Forest model is 99.91148644726914


In [30]:
#Check F-1 score of Random Forest model
print('F1 score of the Random Forest model is {}'.format(f1_score(y_test, rf_yhat)*100))

F1 score of the Random Forest model is 71.6279069767442


In [31]:
#XGBoost
xgb = XGBClassifier(max_depth = 4)
xgb.fit(X_train, y_train)
xgb_yhat = xgb.predict(X_test)

In [33]:
#Check accuracy of XGBoost model
print('Accuracy score of the XGBoost model is {}'.format(accuracy_score(y_test, xgb_yhat)*100))

Accuracy score of the XGBoost model is 99.9506645771664


In [34]:
#Check F-1 score of XGBoost model
print('F1 score of the XGBoost model is {}'.format(f1_score(y_test, xgb_yhat)*100))

F1 score of the XGBoost model is 84.95575221238937


XGBoost model is the most accurate model with a score of 99.9506645771664. 